### PR TITLE
back-end changes for versioning batch uploads

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## Links to jira tickets

- https://quartech.atlassian.net/browse/BHBC-1229

## This PR contains the following changes

- When uploading an occurrence submission, the record is stored in the occurrence_submission table and the file is uploaded to S3 under the path `projects/projectId/surveys/surveyId/submissions/submissionId/filename`

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [x] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [x] I have performed a self-review of my own code

### Tests

- [x] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [x] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

File: database/src/migrations/20210726115350_occurrence_submission_addFilename.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @param {Knex} knex
+ * @return {*}  {Promise<void>}
+ */
+export async function down(knex: Knex): Promise<void> {
+  await knex.raw(`
+    SET SCHEMA '${DB_SCHEMA}';
+    SET SEARCH_PATH = ${DB_SCHEMA},public, biohub_dapi_v1;
+
+    ALTER TABLE occurrence_submission DROP COLUMN file_name;
+
```
Comment (NickPhura) [inline database/src/migrations/20210726115350_occurrence_submission_addFilename.ts:44]: After dropping the column, we need to update the views again.

File: api/src/queries/survey/survey-occurrence-queries.ts [hunk 2]
```diff
@@ -167,8 +207,7 @@ export const getSurveyTemplateS3KeySQL = (surveyId: number, templateId: number):
   }
 
   const sqlStatement: SQLStatement = SQL`
-    SELECT
-      key
+    SELECT *
```
Comment (NickPhura) [inline api/src/queries/survey/survey-occurrence-queries.ts:212]: This function's name is outdated now that the sql is returning the whole row and not just the key.
Same with the JSdoc description and the log labels.

Comment (NickPhura) [reply api/src/queries/survey/survey-occurrence-queries.ts:212]: Maybe call this something like `getOccurrenceSubmissionSQL`

Comment (NickPhura) [reply api/src/queries/survey/survey-occurrence-queries.ts:212]: Also, `surveyId` can be removed from the params/null check as its not used anywhere in the query

Comment (NickPhura) [reply api/src/queries/survey/survey-occurrence-queries.ts:212]: I would also change `tempalteId` to `occurrenceSubmissionId`

File: api/src/paths/project/{projectId}/survey/{surveyId}/template/upload.ts [hunk 3]
```diff
@@ -154,10 +176,10 @@ export function uploadMedia(): RequestHandler {
 export const insertSurveyOccurrenceSubmission = async (
   surveyId: number,
   source: string,
-  key: string,
+  file_name: string,
   connection: IDBConnection
-): Promise<void> => {
-  const insertSqlStatement = insertSurveyOccurrenceSubmissionSQL(surveyId, source, key);
+): Promise<void | any> => {
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/template/upload.ts]: `void | any` may as well just be `any`

File: api/src/paths/project/{projectId}/survey/{surveyId}/template/{templateId}/getSignedUrl.ts [hunk 4]
```diff
@@ -77,11 +77,25 @@ export function getSingleTemplateURL(): RequestHandler {
         getSurveyTemplateS3KeySQLStatement.values
       );
 
+      console.log('result is: ', result);
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/template/{templateId}/getSignedUrl.ts]: remove leftover console

File: api/src/paths/project/{projectId}/survey/{surveyId}/template/{templateId}/getSignedUrl.ts [hunk 5]
```diff
@@ -77,11 +77,25 @@ export function getSingleTemplateURL(): RequestHandler {
...
+        const originalName = result.rows[0].file_name;
+        const templateId = result.rows[0].occurrence_submission_id;
 
-      const s3SignedUrl = await getS3SignedURL(s3Key);
+        const s3Key = generateS3FileKey({
+          projectId: Number(req.params.projectId),
+          surveyId: Number(req.params.surveyId),
+          folder: `submissions/${templateId}`,
+          fileName: originalName
+        });
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/template/{templateId}/getSignedUrl.ts]: We no longer need to generate the key, now that we store the full key in the database.

File: api/src/queries/occurrence/template-observation-queries.ts [hunk 6]
```diff
@@ -22,6 +22,7 @@ export const getTemplateObservationsSQL = (surveyId: number): SQLStatement | nul
     os.update_date,
     os.create_date,
     os.key,
+    os.file_name,
```
Comment (NickPhura) [inline api/src/queries/occurrence/template-observation-queries.ts]: I think this `getTemplateObservationsSQL` function is a duplicate of `getLatestSurveyOccurrenceSubmission`?

File: database/src/migrations/20210726115350_occurrence_submission_addFilename.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    set role biohub_api;
+
+    drop view if exists occurrence_submission;
```
Comment (NickPhura) [inline database/src/migrations/20210726115350_occurrence_submission_addFilename.ts]: We dont want to drop it, but to re-generate it, so the view is back to what it was before this migration ran.

```suggestion
    create or replace view occurrence_submission as select * from biohub.occurrence_submission;
```

File: api/src/queries/occurrence/template-observation-queries.ts (hunk 6 above)
Comment (anissa-agahchen) [reply api/src/queries/occurrence/template-observation-queries.ts]: very close ... I removed the getTemplateObservationSQL and added the pieces I needed to the other function.

**Review by NickPhura — APPROVED** [review]

**Pull request merged**